src/App.tsx

```diff
@@ -15,36 +16,110 @@ export const goodsFromServer = [
+  let visibleGoods = [...goods];
+
+  if (isReversed) {
+    visibleGoods = visibleGoods.toReversed();
+  }
+
+  if (sortBy) {
+    switch (sortBy) {
+      case ESortType.ALPHABETICALLY:
+        visibleGoods = visibleGoods.toSorted();
+        if (isReversed) {
+          visibleGoods = visibleGoods.toReversed();
+        }
+
+        break;
+      case ESortType.LENGTH:
+        if (isReversed) {
+          visibleGoods = visibleGoods.toSorted((a, b) => b.length - a.length);
+        } else {
+          visibleGoods = visibleGoods.toSorted((a, b) => a.length - b.length);
+        }
+
+        break;
+      default:
+        return null;
+    }
+  }
+
+  return visibleGoods;
+};
+
+export const App = () => {
+  const [sortBy, setSortBy] = useState<ESortType>(ESortType.DEFAULT);
+  const [isReversed, setIsReversed] = useState(false);
+  const canReset = sortBy || isReversed;
+
+  const visibleGoods = getGoods(goodsFromServer, { sortBy, isReversed });
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          type="button"
+          className={classNames('button is-info', {
+            'is-light': sortBy !== ESortType.ALPHABETICALLY,
+          })}
+          onClick={() => setSortBy(ESortType.ALPHABETICALLY)}
+        >
           Sort alphabetically
         </button>
 
-        <button type="button" className="button is-success is-light">
+        <button
+          type="button"
+          className={classNames('button is-success', {
+            'is-light': sortBy !== ESortType.LENGTH,
+          })}
+          onClick={() => setSortBy(ESortType.LENGTH)}
+        >
           Sort by length
         </button>
 
-        <button type="button" className="button is-warning is-light">
+        <button
+          type="button"
+          className={classNames('button is-warning', {
+            'is-light': !isReversed,
+          })}
+          onClick={() => setIsReversed(!isReversed)}
+        >
           Reverse
         </button>
 
-        <button type="button" className="button is-danger is-light">
-          Reset
-        </button>
+        {canReset ? (
+          <button
+            type="button"
+            className="button is-danger is-light"
+            onClick={() => {
+              setSortBy(ESortType.DEFAULT);
+              setIsReversed(false);
+            }}
+          >
+            Reset
+          </button>
+        ) : null}
       </div>
 
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        {visibleGoods &&
+          visibleGoods.map(good => (
+            <li key={good} data-cy="Good">
+              {good}
+            </li>
+          ))}
```

Comment:
You can assign an empty array to visibleGoods variable right away so you won't need to take care of it in the rendering logic

```suggestion
     const visibleGoods = getGoods(props) || []; // if null or undefined it will be empty array
    
      return (
          ...some JSX stuff
         {visibleGoods.map(good => ( // if array is empty, there will be nothing rendered
            <li key={good} data-cy="Good">
              {good}
            </li>
          ))}
```

Reply:
thanks, every comment was fixed